### DISCUSSION TITLE
Welcome to discussions!

### DISCUSSION THREAD

**Discussion by maxkeppeler** [maxkeppeler/sheets#12] [General]
## 👋 Welcome!
Here you can: 
  * ❓Ask questions you’re wondering about.  Search the knowledge base as well as existing discussions first.
  * 💡 Share ideas.
  * 📱 Showcase your implementation of this library. They will be mentioned [here](https://github.com/MaxKeppeler/bottom-sheets#showcase) if you like. 
  * ⚠️ If you found a bug, create an issue.
  * ⚒️ If you have a concrete feature request, create an issue for it.

**Comment by irvine752** [comment]
Hey, a couple of questions. These are not issues, just observations I just noticed.

If you don't add the "onPositive" or "onNegative" to an input sheet, the "onResult" is not called. Is this the intended behavior? I simply just had an InputSheet with a list of InputCheckBoxes.

For the button styles, only the "ButtonStyle.OUTLINE" shows the icon. It seems like the "ButtonStyle.NORMAL" leaves some space for an icon, but actually never adds the icon.

**Reply by maxkeppeler** [reply]
Hi! Do you mean the result listener for an individual input or all input data when the user clicks the positive button?
By setting up the positive button with a listener, the result data (one or all inputs) are returned within it. If you don't set up a positive button there's no way you can know the result data (of all inputs). The setup of a negative button has no influence on this however.

I noticed that when I focus on the button (ButtonStyle.NORMAL) the icon color has the same color as the background of the button. The color of the icon should match the color of the text of the button. If it has no focus, it looks like there's space for the icon but no icon. Can you check if this is the case for you too?

**Reply by irvine752** [reply]
Thanks for the quick response. I had individual result listeners for each of the individual inputs. I had to create the onPositive to get each of those called. Your explanation makes sense. I guess the default buttons provided don't really do much besides just closing the sheet. Maybe you could force the behavior by not providing the default ok & cancel buttons for the InputSheet? That would force people to create the positive button.

For the button, very interesting case. If I also put focus on the button, it does show the icon, and you're right the icon color matches the background of the button. Switching the color to match the text color should fix it. 

Side note, I think you should have me as a designated tester for all your projects :) It's always interesting to see what people are working on. If I haven't mentioned this already, great work with this project. It's well designed.

**Reply by maxkeppeler** [reply]
I actually was just implementing it in another project and I used the individual result listeners for changing the values of an object and used the result listener of the sheet as an action to update the model in the database. 
I think I will change it and invoke the individual result listeners, no matter if the sheet result listener was set or not.

**Reply by maxkeppeler** [reply]
I don't really want to force a developer to follow a specific behavior, because I might not think about all use-cases. Though at worst it's just an additional "decoration" for the user. 

What I might look into though, is to perhaps hide the buttons view if the setup configuration of a sheet type allows it (or doesn't enforce it.) I did something with the "toolbar" of the sheet as well. If nothing is set up (Title, IconButtons, …) it is hidden by default.

**Reply by maxkeppeler** [reply]
I fixed the button icon color bug. (Created an issue out of this conversation.) Thanks!

**Reply by maxkeppeler** [reply]
You leave by far the most feedback and lead to lots of bug fixes. :) 
I'm very thankful to you, feel free to help to test my projects! 
I'm aware as well that I should add automated tests, but I didn't find the time to start that (or learn about it) yet.

**Reply by maxkeppeler** [reply]
I will publish 2.2.0 tomorrow noon, in case you notice that something isn't working, let me know before!

**Comment by irvine752** [comment]
Not sure if you have this documented somewhere, but what's the best way to use some of the Input Options in a custom sheet?

**Reply by maxkeppeler** [reply]
Hey, what do you mean exactly? 
You can use all the input types from the package ´´´com.maxkeppeler.sheets.input.type```.
```
  InputSheet().show(...) {
            title("Some form")
            with(InputEditText {
                ...
            })
            with(InputCustom {
                ...
            })
           with(InputSeparator {
                ...
            })
           with(InputSwitch {
                ...
            })
           with(InputButtonToggleGroup {
                ...
            })
            with(InputCheckBox("some_key") { 
                ...
            })
            with(InputSpinner {
                ...
            })
            with(InputRadioButtons("") {
               ...
            })
            onNegative { 
             }
            onPositive { result ->
                // read values by result bundle if preferred
            }
        }
```

**Reply by maxkeppeler** [reply]
Oh you mean using the input types and configurations in self-created sheet? The InputSheet exists to give a quick and powerful way to make forms. You can not use them in another sheet or your own sheet implementation.  If you inherit from Sheet as it is written in the README, you can however use your own layout and components to build a sheet to fit your use-cases. (As you would normally do.)

**Reply by irvine752** [reply]
Yep, I meant the input types in the self-created/custom sheets. I was able to extend the Sheet class to use my my own custom layout. Is there any reason the InputSheet class is marked as a final class?

**Reply by maxkeppeler** [reply]
Well, all Sheet-classes are implementations of specific use-cases based on the abstract Sheet-class. One use-case, one final implementation. As the InputSheet has additional logic for visibility, validation and each input type's setup in the adapter, I don't see why it shouldn't be a final class. Could you explain a bit further what you imagine/ want?

**Comment by irvine752** [comment]
Max, is there anyway to add text in HTML format? I was looking to add some text with bullets to the InfoSheet. This would be for the content.

**Reply by maxkeppeler** [reply]
Not yet, if I have some time, I will work on the current issues and your request. Likely in 2 weeks or so. (Really want to make a compose version of this lib as well)

**Comment by paragshedbale** [comment]
Hi Max,

Fantastic work @ sheets.  I have just started using them. I have run into a problem, ignorant me. When I use the CalendarSheet, I do not see a positive button. And the selected date range at the top left corner only shows day and month. What should I do to eliminate these issues?

**Reply by maxkeppeler** [reply]
Hey! The positive button appears at most sheets only once the input is valid. In the case of CalendarSheet, if a range or date was selected. You are right, the selected date range shows only the day and month. The month and year selection on the right side only applies to the current calendar view, however the user can see which year the range lays. The only problem I see is, if the user selects a range between the end of a year and the beginning of the next year.

**Reply by paragshedbale** [reply]
Thanks very much Max. I think I realised that @ appears only if the selected values are appropriate.

If I select a date range within a week the positive button does indeed appear. But I would like to select a date range for say 3 months. If I do that currently, it seems that for some reason it's stipulated as non valid date range, since the positive button doesn't appear. How do I make that happen, ie. select a date range of say the last 67 days?

Thanks once again.

**Reply by paragshedbale** [reply]
Setting max date and min date range? Haven't tried that. Will try and get back. Thanks.
